### ISSUE TITLE
Ignore duplicate variants of same migration (e.g. `.js`/`.ts`)

### ISSUE THREAD

**Issue by sgoll** [tgriesser/knex#2756]
# Environment

Knex version: 0.15.2
Database + version: any
OS: any

# Feature discussion / request

## Use case

Code that implements migrations in TypeScript (or CoffeeScript or some other dialect that gets compiled to plain JavaScript) needs to be aware of different filename extensions in development VS production.

Even with the `loadExtensions` setting introduced in #2168, it is sometimes difficult to tell exactly when 
to use (only) the `.ts` file and when to use (only) the `.js` file.

One particular issue that I experience is this:
- During development, I use `ts-node` to run Knex migrations. I want that to use the `.ts` files.
- During testing, I rely on Jest and Wallaby.js. Jest sees only the `.ts` files. Wallaby.js sees the `.ts` files and `.js` files created just-in-time directly next to them. I want both to use the `.ts` files only.
- During production, I only have the `.js` files available in the migrations directory. Nonetheless, I want to be able to operate on the same database instance that I used during development, i.e. seamlessly run the dev version or the prod version of my application on any database.
- Ideally, I would use the same Knex configuration and not adjust more settings than necessary between dev and prod to make tests more reliable.

## Requested feature

Use `loadExtensions` to only load any given basename once, with the extension that first appears in the `loadExtensions` list.

Example: Suppose the migrations directory contains the following files:
- `001.ts` (A, original TypeScript), `001.js` (A, built file)
- `002.ts` (B, original TypeScript)
- `003.js` (C, built file)

We have three migrations (`001`, `002`, `003`). When `loadExtensions` is set to `[".ts", ".js"]`, Knex should only consider the following files to apply as migrations instead of all four files as is the case now:
- `001.ts` (`.ts` has priority over `.js` as it is listed before the latter in `loadExtensions`)
- `002.ts` (there is no other file for the `002` basename)
- `003.js` (ditto)

## API / Implementation proposal

Function `listAll()` in `src/migrate/migration-list-resolver.js` could be adjusted to only return the first instance of each migration with the same basename (i.e. migrations that only differ in their extension), with priority of extension given by their order in `loadExtensions` as described above.

In addition, it would be necessary to replace calls to lodash's `difference()` with `differenceWith()` so that extensions get ignored when comparing all available migrations to the ones that have already been applied and stored in the database (returned by `listAll()` and `listCompleted()`, respectively).

Alternatively, it would be possible to only ever operate on the basenames of each migration file and store only that in the database (i.e. a migration would only refer to the basename instead of the complete filename). This would, however, be a breaking change and probably require a flag in the config to activate this feature, though it would arguably be the cleaner way of implementing the suggested feature.

Maybe it would be possible to implement the suggested changes as a follow-up to the support for multiple directories in #2690 (#2735).

## Possible side-effects

Migrations that share the same basename but have different extensions would be seen as a single migration when this proposal gets implemented. Previously, every file in the migrations directory would be considered a migration. If one had deliberately chosen two migrations to have the same basename and differ only in their extension, only one of them would get applied with this proposal implemented.

However, since it is bad practice to have two different modules with the same basename and different extensions, this should not be a typical use case. This proposal is therefore not expected to break any existing migrations.

NB: All things said in this issue would apply to both migrations as well as seed files.

**Comment by kibertoad** [comment]
Should be a pretty straightforward change; I would be happy to implement it should it be approved (and after #2690 is merged :-P ). Considering that this would be a breaking change, probably it should be an opt-in behaviour.

**Comment by elhigu** [comment]
I'm fine with doing this as breaking change and just release it in 0.16

**Comment by flipace** [comment]
We have similar issues here, and we were thinking that it would actually be enough for us if it were possible to ignore the extension (omitting it when storing the filename in the database or ignoring it when comparing the list of migrations).

**Comment by kibertoad** [comment]
I'll pick this task up after https://github.com/tgriesser/knex/pull/2775 is merged, otherwise I might introduce some conflicts.

**Comment by iilei** [comment]
I am about to start working on a patch that allows pattern matching utilizing [globby](https://www.npmjs.com/package/globby). It seems to me that it would also have potenial for merge conflicts, so my question is @kibertoad : as the mentioned PR is merged – is it a valid assumtion it is better to forge my patch on top of this branch?

**Comment by iilei** [comment]
@sgoll Assuming in your usage scenario the compiled files are to be ignored from version control I suppose it would be also an option to have them ignored via a pattern as it is provided to `.gitignore` - right?

If so, the work I am doing on my patch mentioned above would provide an interface to globby - [allowing to ignore the same files based on `.gitignore`](https://www.npmjs.com/package/globby#globbygitignoreoptions)

**Comment by kibertoad** [comment]
@iilei All of my changes were merged to master already, so feel free to branch off that.

**Comment by sgoll** [comment]
@iilei

> Assuming in your usage scenario the compiled files are to be ignored from version control I suppose it would be also an option to have them ignored via a pattern as it is provided to `.gitignore` - right?

Yes, that it correct: the compiled files are not put into version control (in my case).

> If so, the work I am doing on my patch mentioned above would provide an interface to globby - [allowing to ignore the same files based on `.gitignore`](https://www.npmjs.com/package/globby#globbygitignoreoptions)

That would be quite useful. However, my original intention here was not to unconditionally ignore `.js` files but tell Knex that it should prefer `.ts` over `.js` files when both are available, and use either in case only one is available (see rationale with regard to testing versus production environment in the original post).

I am not sure how that would work with your proposed changes. Can you elaborate on that?

**Comment by elhigu** [comment]
But sholdnt it work other way around, that compiled/.gitignore files are actually ran 🤔 doesnt make much sense to compile them if they will never be ran...

**Comment by elhigu** [comment]
That sounds a bit too opinionated behaviour to my taste.

**Comment by sgoll** [comment]
@elhigu 

> But sholdnt it work other way around, that compiled/.gitignore files are actually ran  doesnt make much sense to compile them if they will never be ran...

Jest uses the raw `.ts` files for running tests. This is particularly useful when running Jest in `--watch` mode for test-driven development.

**Comment by elhigu** [comment]
@sgoll I understand that use case very well (along with others), but it doesnt really invalidate my comment.

**Comment by JakeGinnivan** [comment]
@iilei you mean like https://github.com/knex/documentation/pull/133?

**Comment by sgoll** [comment]
@elhigu 

> I understand that use case very well (along with others), but it doesnt really invalidate my comment.

You are right (and I never intended to insinuate that you weren't). Personally, I think an approach like the one suggested by @iilei would work best: have the ability to resolve migrations (and seed files) with some custom resolver function.

This resolver function would be able to support both use cases:

- Select `.ts` over `.js` when both are present
- Implement elaborate glob pattern matching

The issue that remains is how to name the resolved migrations when they are put into the database (so that both `.ts` and `.js` end up with the same unique name). But that could also be part of the resolver function, maybe returning the extra data along with the raw filenames, e.g.

```js
resolveMigrations: function () {
  return [
    { path: './migrations/001.ts', name: '001' },
    { path: './migrations/002.js', name: '002' },
  ];
}
```

**Comment by JakeGinnivan** [comment]
@sgoll not sure if you had a look at my links, but this functionality exists already inside knex `@next`. 

```
     class MyMigrationSource {
        // Must return a Promise containing a list of migrations. 
        // Migrations can be whatever you want, they will be passed as
        // arguments to getMigrationName and getMigration
        getMigrations() {
          // In this example we are just returning migration names
          return Promise.resolve(['migration1'])
        }
        getMigrationName(migration) {
          return migration;
        }
        getMigration(migration) {
          switch(migration) {
            case 'migration1':
              return {
                up(knex)   { /* ... * / }
                down(knex) { /* ... * / }
              }
          }
        },
      }
      // pass an instance of your migration source as knex config
      knex.migrations.latest({ migrationSource: new MyMigrationSource() })
```

**Comment by sgoll** [comment]
@JakeGinnivan 

> not sure if you had a look at my links, but this functionality exists already inside knex `@next`.

That looks exactly like what I need. I'll give it a try as soon as possible.

Is there something similar for seed files as well?

**Comment by JakeGinnivan** [comment]
I didn't do seed files, that would be a good enhancement.

**Comment by tejasmanohar** [comment]
Is there a canonical solution here? I renamed a migration file from .js to .ts when migrating to Typescript, and I think Knex thinks it needs to re run the migration

**Comment by elhigu** [comment]
@tejasmanohar easiest way for you is to manually fix filenames in migration table. I don't thing that separating extension from migration filename check has been done.

**Comment by tejasmanohar** [comment]
@elhigu Cool cool, thats what I figured. Thanks!

**Comment by mkalam-alami** [comment]
Leaving my own approach here, which I've been using **for launching migrations either from a TS or JS server**. This is required for me as I launch migrations from code at launch time, and the server may be either in TS or JS depending on the environment (TS = dev time using ts-node, JS = production build).

```typescript
if (__filename.endsWith(".js")) {
  await knex(migrationsTableName).update({
    name: knex.raw("REPLACE(name, '.ts', '.js')")
  });
} else {
  await knex(migrationsTableName).update({
    name: knex.raw("REPLACE(name, '.js', '.ts')")
  });
}
```

**Comment by TrueWill** [comment]
👍 for this - we just encountered the same issue. We were transpiling to .js but then switched to ts-node; now we're getting
> Error: The migration directory is corrupt

because the files end in .ts instead of .js.

**Comment by cerino-ligutom** [comment]
Any updates on this?

**Comment by mercmobily** [comment]
Thank you for this well-thought-out feature request about handling duplicate migration variants with different extensions. This is indeed a common pain point for TypeScript users, and your analysis of the problem is spot-on.

**The Problem**

When using TypeScript for migrations, developers often end up with both `.ts` source files and `.js` compiled files in the same directory. Knex treats these as separate migrations, leading to errors like "The migration directory is corrupt" when switching between development (using `.ts` files) and production (using `.js` files).

This issue has affected many users since 2018, with people still encountering it in 2020 and beyond, despite various workarounds being available.

**Proposed Solution**

Your original proposal remains the best approach and should be implemented in Knex core:

1. **Use `loadExtensions` order as priority**
   - When multiple files share the same basename (e.g., `001.ts` and `001.js`), Knex should use only the first matching extension from `loadExtensions`
   - Example: With `loadExtensions: ['.ts', '.js']`, if both `001.ts` and `001.js` exist, only `001.ts` would be used

2. **Store migrations without extensions in the database**
   - The migration table should store `001` instead of `001.ts` or `001.js`
   - This allows seamless switching between TypeScript development and JavaScript production environments
   - Existing migrations with extensions would need migration logic for backward compatibility

3. **Configuration approach**
   - Add a new option: `migrations: { ignoreExtensionDuplicates: true }`
   - When enabled, implements the behaviour described above
   - Could potentially become the default behaviour in a future major version

**Implementation Details**

The changes would primarily affect:
- `listAll()` in migration-list-resolver.js - to filter duplicates based on `loadExtensions` priority
- Migration comparison logic - to use `differenceWith()` instead of `difference()` for extension-agnostic comparison
- Migration storage - to strip extensions before storing in the database
- Migration lookup - to match basenames regardless of extension

This solution would:
- Work out of the box for the common TypeScript use case
- Maintain backward compatibility (as an opt-in feature initially)
- Eliminate the need for complex workarounds
- Follow the principle of least surprise

**Why This Hasn't Been Implemented Yet**

Despite being a common problem with a clear solution:
- Concerns about breaking changes for users with same-named files intentionally
- Lack of consensus on whether it should be opt-in or default behaviour
- The existence of workarounds (though they're not ideal) reduced urgency

Given the growth of TypeScript since 2018, this deserves priority as a core feature.

**Available Workarounds** (until this is implemented)

While we wait for the proper solution, here are the current workarounds:

**Option 1: Custom Migration Source** (available since v0.16.0, but requires significant boilerplate)

```javascript
const path = require('path');
const fs = require('fs').promises;

class TypeScriptPreferringMigrationSource {
  constructor(migrationDirectory, loadExtensions = ['.ts', '.js']) {
    this.migrationDirectory = migrationDirectory;
    this.loadExtensions = loadExtensions;
  }

  async getMigrations() {
    const files = await fs.readdir(this.migrationDirectory);
    
    // Group files by basename
    const migrationsByBasename = {};
    for (const file of files) {
      const ext = path.extname(file);
      if (this.loadExtensions.includes(ext)) {
        const basename = path.basename(file, ext);
        if (!migrationsByBasename[basename]) {
          migrationsByBasename[basename] = [];
        }
        migrationsByBasename[basename].push({ file, ext });
      }
    }
    
    // Select preferred extension for each basename
    const selectedMigrations = [];
    for (const [basename, variants] of Object.entries(migrationsByBasename)) {
      variants.sort((a, b) => {
        return this.loadExtensions.indexOf(a.ext) - this.loadExtensions.indexOf(b.ext);
      });
      selectedMigrations.push({
        file: variants[0].file,
        name: basename // Store without extension
      });
    }
    
    return selectedMigrations.sort((a, b) => a.name.localeCompare(b.name));
  }

  getMigrationName(migration) {
    return migration.name;
  }

  async getMigration(migration) {
    const fullPath = path.join(this.migrationDirectory, migration.file);
    return require(fullPath);
  }
}
```

**Option 2: Dynamic Database Updates** (@mkalam-alami's approach)

```javascript
if (__filename.endsWith(".js")) {
  await knex(migrationsTableName).update({
    name: knex.raw("REPLACE(name, '.ts', '.js')")
  });
} else {
  await knex(migrationsTableName).update({
    name: knex.raw("REPLACE(name, '.js', '.ts')")
  });
}
```

**Option 3: Manual Migration Table Updates**

Directly update the migration table when switching between TypeScript and JavaScript environments.

These workarounds demonstrate that the solution is well-understood - it just needs to be built into Knex core to provide a good developer experience.

**Related issues:**

- #3849 - JavaScript knexfile can't run TypeScript migrations  
- #3019 - loadExtensions works incorrectly
- #3867 - Migration directory error when using Custom Migration Source
- #5798 - Ignore the FS-related option for loadExtensions when providing custom migrationSource
- #5800 - Proper support for TypeScript and ESM (related but doesn't solve this specific issue)

**Comment by mercmobily** [comment]
I am marking it as a bug, since it effectively crashes when it shouldn't

**Comment by mercmobily** [comment]
I want to add that there is an existing PR created for an issue that is now marked as duplicate of this one: #4802 It should be looked at
